[nativert] move recordfunction

dolpm opened the pull request:
Summary:
nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff moves the our function-recording raii wrapper into record_function_ops

Test Plan: CI

Differential Revision: D74284301

albanD commented on torch/csrc/autograd/record_function_ops.h:
```diff
@@ -24,4 +26,29 @@ TORCH_API c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut_new
     const c10::intrusive_ptr<c10::ivalue::Future>& fut);
 
+/**
+ * RAII wrapper that behaves similarly to torch.profiler.record_function.
```
The profiler API doesn't have such API already?

albanD commented on torch/csrc/autograd/record_function_ops.h:
```diff
@@ -24,4 +26,29 @@ TORCH_API c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut_new
+/**
+ * RAII wrapper that behaves similarly to torch.profiler.record_function.
+ */
+class RecordFunction {
```
Isn't this part of libtorch_python? Would nativert depend on libtorch-only or libtorch_python?

albanD replied:
Ok, let's add a test to make sure all of this behave as expected

cyyever commented on torch/csrc/autograd/record_function_ops.h:
```diff
@@ -24,4 +26,29 @@ TORCH_API c10::intrusive_ptr<c10::ivalue::Future> _call_end_callbacks_on_fut_new
+class RecordFunction {
+ public:
+  RecordFunction() = delete;
+  RecordFunction(const RecordFunction&) = default;
```
These move and copy constructors should be `delete`.

dolpm committed:
[nativert] move recordfunction (#153088)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/153088

nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff moves the our function-recording raii wrapper into record_function_ops

Test Plan: CI

Differential Revision: D74284301

jeanschmidt commented:
Merged internally but no valid OSS approval, closing this PR and jedi unlanding [D74284301](https://www.internalfb.com/diff/D74284301)
